### PR TITLE
platform: imx: Add support for record with the ESAI

### PR DESCRIPTION
**Pull request by paulstelian97**
The ESAI supports both capture and playback and also to use them simultaneously.
This patch set enables that functionality.

### CODE REVIEW

**Comment by paulstelian97** [comment]
@tlauda This is where I have attempted to add said support. As I mentioned in the mail on the mailing list I'm getting a -512 size in the DMA SG elems, and I believe the function host_dma_get_split is miscomputing splits when `bytes` is 0, leading to a negative size overall.

**Comment by paulstelian97** [comment]
To report: There is some functionality that I already have (if I manage to synchronize aplay and arecord well I can pipe from one to the other or run them in parallel but that's about coincidence for now; playback and record work individually).

@groncarolo @jlacla Does one of you know enough about either the ESAI or the cs42888 codec to help me? Perhaps there is a way to use the SYN bit from SAICR? (I have tried and got some really bad quality on the recording side consistently)

**Comment by dbaluta** [comment]
@paulstelian97 we need to talk with @TE-N-ShengjiuWang to understand if we need to support playback / record in parallel on the same DAI.

@tlauda @jajanusz do you guys support aplay & arecord working in the same time on the same sound card?

**Comment by TE-N-ShengjiuWang** [comment]
@paulstelian97  we have support the full duplex without DSP,  with DSP, this feature should also be supported.  you can dump the registers  to compare with no DSP case (not sof), to find a proper register setting.

**Comment by paulstelian97** [comment]
@TE-N-ShengjiuWang I did that and I have noticed that we have a txs-rxm configuration for the codec (the PR as of this writing currently assumes txs-rxs). I will update it but do you have any idea of the technical reasons why it was chosen to be like this in the DTS?

**Comment by TE-N-ShengjiuWang** [comment]
@paulstelian97  it is a workaround for the board issue,  that on mek, if we use txs-rxs or txm-rxm,  one direction ( playback or caputure) will be noise.  the reason is the clock line (bitclock & Frame clock) is too long from codec to imx chip.

why we select the txs-rxm in dts,  it because it is a easy way to do this :),   for the ALSA don't support
set txs-rxm in machine driver for a same hardware device.   the formal way to workaround this issue is to generate two sound card, hw:0,0, and hw:0,1 for example.  hw:0,0 for playback,  hw:0,1 for recording.

**Comment by paulstelian97** [comment]
@TE-N-ShengjiuWang Alright, what the DTS selects will not be propagated so for now I'll just hardcode this particular issue of the configuration. I believe I'll bring an update in up to an hour.

**Comment by paulstelian97** [comment]
OK now it seems to work as it should, thanks for the input. I'll clean up later then mark as ready for review.

File: src/drivers/imx/esai.c [hunk 1]
```diff
@@ -197,17 +197,23 @@ static inline int esai_set_config(struct dai *dai,
 
 	xccr |= ESAI_xCCR_xHCKD; /* Set the HCKT pin as an output */
 
-	dai_update_bits(dai, REG_ESAI_TCCR, mask, xccr); /* rx */
-	dai_update_bits(dai, REG_ESAI_RCCR, mask, xccr); /* tx */
+	dai_update_bits(dai, REG_ESAI_TCCR, mask, xccr);
+	/* We must make our rx clock slave because codec is master
+	 * TODO make this come from topology
+	 */
+	xccr &= ~(ESAI_xCCR_xCKD | ESAI_xCCR_xFSD);
```
Comment (dbaluta) [inline src/drivers/imx/esai.c]: Lets add a proper explanation here: "There is a hardware limitation for using tx/rx in SAI
master. Workaround for this is to use tx master  / rx slave"

File: src/drivers/imx/esai.c [hunk 2]
```diff
@@ -197,17 +197,23 @@ static inline int esai_set_config(struct dai *dai,
...
 		ESAI_xCR_xPR;
 	/* Personal reset, suspend the actual TX/RX for now */
 	xcr |= ESAI_xCR_xPR;
 
-	dai_update_bits(dai, REG_ESAI_TCR, mask, xcr); /* rx */
-	dai_update_bits(dai, REG_ESAI_RCR, mask, xcr); /* tx */
+	dai_update_bits(dai, REG_ESAI_TCR, mask, xcr);
+	/* rx doesn't have any PADC, unset it */
+	xcr &= ~ESAI_xCR_PADC;
+	dai_update_bits(dai, REG_ESAI_RCR, mask, xcr);
```
Comment (dbaluta) [inline src/drivers/imx/esai.c]: Is this necessary? Or its just a cosmetic cleanup?

Comment (paulstelian97) [reply src/drivers/imx/esai.c]: The bit that would be at the same position in RCR as PADC is in TCR is reserved and probably ignored. But I still believe reserved bits should be left alone.

Perhaps a better variant is to remove PADC from the mask itself instead?

Comment (dbaluta) [reply src/drivers/imx/esai.c]: @paulstelian97 sure, but please make it as a separate commit. Also, this is no longer an RFC so just update the PR description to reflect this.

Comment (paulstelian97) [reply src/drivers/imx/esai.c]: Alright, doing that.

File: src/drivers/imx/edma.c [hunk 3]
```diff
@@ -345,6 +345,7 @@ static int edma_set_config(struct dma_chan_data *channel,
 	trace_edma("EDMA: set config");
 
 	channel->is_scheduling_source = config->is_scheduling_source;
+	channel->direction = config->direction;
```
Comment (dbaluta) [inline src/drivers/imx/edma.c]: @paulstelian97 this doesn't seem related to this change. So, please put in another patch.

File: src/drivers/imx/esai.c [hunk 4]
```diff
@@ -194,41 +194,48 @@ static inline int esai_set_config(struct dai *dai,
...
 	mask = ESAI_xCR_xFSL | ESAI_xCR_xFSR | ESAI_xCR_xWA |
 		ESAI_xCR_xMOD_MASK | ESAI_xCR_xSWS_MASK | ESAI_xCR_PADC |
 		ESAI_xCR_xPR;
 	/* Personal reset, suspend the actual TX/RX for now */
 	xcr |= ESAI_xCR_xPR;
 
-	dai_update_bits(dai, REG_ESAI_TCR, mask, xcr); /* rx */
-	dai_update_bits(dai, REG_ESAI_RCR, mask, xcr); /* tx */
+	dai_update_bits(dai, REG_ESAI_TCR, mask, xcr);
+	dai_update_bits(dai, REG_ESAI_RCR, mask, xcr);
```
Comment (dbaluta) [inline src/drivers/imx/esai.c]: Why do you remove the comments tx, rx here? Please if you want to do cleanups make a separate patch.

File: src/drivers/imx/esai.c [hunk 5]
```diff
@@ -242,6 +249,12 @@ static inline int esai_set_config(struct dai *dai,
 			ESAI_xFCR_xWA(24) | ESAI_xFCR_xFWM(96)
 			| ESAI_xFCR_TE(1) | ESAI_xFCR_TIEN);
 
+	dai_update_bits(dai, REG_ESAI_RFCR,
+			ESAI_xFCR_xFR_MASK | ESAI_xFCR_xWA_MASK
+			| ESAI_xFCR_xFWM_MASK | ESAI_xFCR_RE_MASK,
+			ESAI_xFCR_xWA(24) | ESAI_xFCR_xFWM(96)
```
Comment (dbaluta) [inline src/drivers/imx/esai.c]: Is it possible to leave | at the end of the line, so that the beginning of the lines are easier to read?

File: src/drivers/imx/esai.c (hunk 4 above)
Comment (paulstelian97) [reply src/drivers/imx/esai.c]: The comments were wrong as far as I understand right now, so I thought of removing them because they were misleading.

**Comment by dbaluta** [comment]
SOFCI TEST

**Comment by paulstelian97** [comment]
On-device test errors unrelated to PR, CamelCase errors necessary due to how the bits/masks/everything are defined in the ESAI, the topology commit doesn't truly need an extended description in my opinion (is there anything extra I can say about how I modified the topology to support capture?).

**Comment by dbaluta** [comment]
>  in the ESAI, the topology commit doesn't truly need an extended description in my opinion (is there anything extra I can say about how I modified the topology to support capture?).

No need for extra description on that.

**Comment by dbaluta** [comment]
@groncarolo @jlacla care to have a look?

**Comment by paulstelian97** [comment]
Rebased (was like 39 commits behind) + minor reword (record -> capture) in one of the commit messages.

**Comment by lgirdwood** [comment]
CI jenkins unrelated.
